### ISSUE TITLE
mix compile fails

### ISSUE THREAD

**Issue by IRog** [processone/ejabberd#1774]
> What version of ejabberd are you using?

Head/Master

> What operating system (version) are you using?

Mac and Ubuntu

> How did you install ejabberd (source, package, distribution)?

Source

> What did not work as expected? Are there error messages in the log? What
> was the unexpected behavior? What was the expected result?

`mix compile` with Erlang 19.3 and Elixir 1.4.4 fails when run in the project root. I tried using Elixir 1.3.2, as well.

```
$mix compile
Compiling 2 files (.erl)
src/mod_muc_room.erl:3691: field jid undefined in record muc_subscribe
src/mod_muc_room.erl:3696: variable 'SubJid' is unbound
src/mod_muc_room.erl:3698: field jid undefined in record muc_subscribe
src/mod_muc_room.erl:3735: field jid undefined in record muc_unsubscribe
src/mod_muc_room.erl:3740: variable 'UnsubJid' is unbound
src/mod_muc_room.erl:3742: field jid undefined in record muc_unsubscribe
src/mod_http_upload.erl:627: undefined macro 'NS_HTTP_UPLOAD_0'
src/mod_http_upload.erl:688: undefined macro 'NS_HTTP_UPLOAD_0'
src/mod_http_upload.erl:473: function iq_disco_info/4 undefined
src/mod_http_upload.erl:482: record upload_request_0 undefined
src/mod_http_upload.erl:487: variable 'CType' is unbound
src/mod_http_upload.erl:487: variable 'File' is unbound
src/mod_http_upload.erl:487: variable 'Size' is unbound
src/mod_http_upload.erl:487: variable 'XMLNS' is unbound
src/mod_http_upload.erl:510: function mk_slot/3 undefined
src/mod_http_upload.erl:513: function mk_slot/3 undefined
src/mod_http_upload.erl:620: spec for undefined function mk_slot/3
src/mod_http_upload.erl:670: spec for undefined function iq_disco_info/4
** (Mix) Encountered compilation errors
```

This is keeping me from embedding ejabberd in an Elixir application, as the `mix deps.compile ejabberd` fails in that context and this is a better example of the problem.

**Comment by IRog** [comment]
Here is the error from `mix deps.compile ejabberd`

```
src/ejabberd_s2s_in.erl:24: Warning: behaviour ejabberd_socket undefined
src/ejabberd_s2s_in.erl:311: Warning: function check_from_to/3 is unused
src/ejabberd_s2s_in.erl:325: Warning: function check_from/2 is unused
src/ejabberd_service.erl:40: can't find include file "xmpp.hrl"
src/ejabberd_service.erl:105: undefined macro 'NS_COMPONENT'
src/ejabberd_service.erl:181: undefined macro 'is_stanza/1'
src/ejabberd_service.erl:33: function init/1 undefined
src/ejabberd_service.erl:34: function handle_authenticated_packet/2 undefined
src/ejabberd_service.erl:61: type xmpp_element() undefined
src/ejabberd_service.erl:228: type jid() undefined
src/ejabberd_service.erl:237: record jid undefined
src/ejabberd_service.erl:23: Warning: behaviour xmpp_stream_in undefined
src/ejabberd_service.erl:24: Warning: behaviour ejabberd_socket undefined
src/ejabberd_service.erl:229: Warning: function check_from/2 is unused
src/ejabberd_service.erl:240: Warning: function random_password/0 is unused
could not compile dependency :ejabberd, "mix compile" failed. You can recompile this dependency with "mix deps.compile ejabberd", update it with "mix deps.update ejabberd" or clean it with "mix deps.clean ejabberd"
==> ejapp
** (Mix) Encountered compilation errors
```

**Comment by sformanjawbone** [comment]
I'm seeing this when trying to build using the Docker image https://hub.docker.com/r/ejabberd/mix/ (as per the blog post https://blog.process-one.net/ejabberd-development-with-docker/ )

The commands I used are:

1. `docker pull ejabberd/mix`
2. `git clone https://github.com/processone/ejabberd.git`
3. `cd ejabberd/` (this command is unmentioned in the blog post)
4. `docker run --rm -v $(pwd):$(pwd) -w $(pwd) ejabberd/mix do deps.get, deps.compile, compile`

and the result (the attached file has a full trace of my shell session):

```
[ProblemBuildingEjabberdDocker.txt](https://github.com/processone/ejabberd/files/1069293/ProblemBuildingEjabberdDocker.txt)
==> ejabberd
Compiling 1 file (.asn1)
Compiling 1 file (.yrl)
Compiling 242 files (.erl)
src/mod_pres_counter.erl:28: Warning: behaviour gen_mod undefined
src/mod_ping.erl:32: Warning: behaviour gen_mod undefined
src/ejabberd_access_permissions.erl:32: Warning: behaviour ejabberd_config undefined
src/mod_client_state.erl:31: Warning: behaviour gen_mod undefined
src/mod_muc_room.erl:3691: field jid undefined in record muc_subscribe
src/mod_muc_room.erl:3696: variable 'SubJid' is unbound
src/mod_muc_room.erl:3698: field jid undefined in record muc_subscribe
src/mod_muc_room.erl:3735: field jid undefined in record muc_unsubscribe
src/mod_muc_room.erl:3740: variable 'UnsubJid' is unbound
src/mod_muc_room.erl:3742: field jid undefined in record muc_unsubscribe
src/randoms.erl:53: Warning: crypto:rand_bytes/1 is deprecated and will be removed in a future release; use crypto:strong_rand_bytes/1
src/mod_http_upload.erl:627: undefined macro 'NS_HTTP_UPLOAD_0'
src/mod_http_upload.erl:688: undefined macro 'NS_HTTP_UPLOAD_0'
src/mod_http_upload.erl:473: function iq_disco_info/4 undefined
src/mod_http_upload.erl:482: record upload_request_0 undefined
src/mod_http_upload.erl:487: variable 'CType' is unbound
src/mod_http_upload.erl:487: variable 'File' is unbound
src/mod_http_upload.erl:487: variable 'Size' is unbound
src/mod_http_upload.erl:487: variable 'XMLNS' is unbound
src/mod_http_upload.erl:510: function mk_slot/3 undefined
src/mod_http_upload.erl:513: function mk_slot/3 undefined
src/mod_http_upload.erl:620: spec for undefined function mk_slot/3
src/mod_http_upload.erl:670: spec for undefined function iq_disco_info/4
src/mod_muc_mnesia.erl:28: Warning: behaviour mod_muc_room undefined
src/mod_muc_sql.erl:30: Warning: behaviour mod_muc_room undefined
src/mod_muc_riak.erl:28: Warning: behaviour mod_muc_room undefined
** (Mix) Encountered compilation errors
```

**Comment by sformanjawbone** [comment]
This is the full trace of my shell session I mentioned in the previous comment.  I thought I had attached it but I don't see it there so I'm adding this additional comment to include it.
[ProblemBuildingEjabberdDocker.txt](https://github.com/processone/ejabberd/files/1069299/ProblemBuildingEjabberdDocker.txt)

**Comment by IRog** [comment]
For reference @jglover Use my branch, processone's master is currently not working because it's relying on some changes in xmpp library that hasn't been released yet (eg this constant processone/xmpp@06d4d57#diff-374df80684a5cc6e2e1080f0b65ddbd0R159)

**Comment by cromain** [comment]
mix compilation was fixed recently, thanks to upgrade of some hex.pm packages
